container/shim/src/fetchers/lassie.js

```diff
@@ -1,3 +1,143 @@
+  }
+
+  const lassieUrl = new URL(LASSIE_ORIGIN + toUtf8(req.path));
+  for (const [key, val] of Object.entries(req.query)) {
+    lassieUrl.searchParams.set(key, toUtf8(val));
+  }
+  lassieUrl.searchParams.set("format", "car");
+
+  const controller = new AbortController();
+  const timeout = setTimeout(() => controller.abort(), LASSIE_TIMEOUT);
+
+  const _http = lassieUrl.protocol === "https:" ? https : http;
+  const agent = lassieUrl.protocol === "https:" ? httpsAgent : httpAgent;
+
+  const lassieReq = _http
+    .get(
+      lassieUrl,
+      {
+        agent,
+        timeout: LASSIE_TIMEOUT,
+        signal: controller.signal,
+      },
+      async (fetchRes) => {
+        clearTimeout(timeout);
+        const { statusCode } = fetchRes;
+
+        res.status(statusCode);
+
+        if (statusCode === 200) {
+          res.set("Cache-Control", "public, max-age=29030400, immutable");
+        } else if (statusCode >= 400) {
+          debug.extend("error")(`Invalid response from Lassie (${statusCode}) for ${cid}`);
+          res.end();
+        }
+
+        if (isRawFormat) {
+          getRequestedBlockFromCar(fetchRes, res).catch((err) => debug(err));
+        } else {
+          proxyResponseHeaders(fetchRes, res);
+          streamCAR(fetchRes, res).catch(() => {});
+        }
+      }
+    )
+    .on("error", (err) => {
+      clearTimeout(timeout);
+      debug.extend("error")(`Error fetching from Lassie for ${cid}: ${err.name} ${err.message}`);
+      if (controller.signal.aborted) {
+        return res.sendStatus(504);
+      }
+      if (!res.headersSent) res.sendStatus(502);
+    })
+    .on("timeout", () => {
+      clearTimeout(timeout);
+      debug.extend("error")(`Timeout from Lassie for ${cid}`);
+      lassieReq.destroy();
+      res.destroy();
+    });
+
+  req.on("close", () => {
+    clearTimeout(timeout);
+    if (!res.writableEnded) {
+      debug.extend("error")("Client aborted early, terminating gateway request");
+      lassieReq.destroy();
+    }
+  });
+}
+
+/**
+ * @param {IncomingMessage || ReadableStream} req
+ * @param {Response} res
+ * @param {string} cid
+ * @param {Uint8Array} block
+ */
+function respondFromBlockCache(req, res, cid, block) {
+  res.status(200);
+  res.set("content-length", Buffer.byteLength(block));
+  res.set("content-type", "application/vnd.ipld.raw");
+  // TODO: Write headers
+  res.end(block);
+}
+
+/**
+ * The first block is returned to the client.
+ * All other blocks (if any) are written to the block cache.
+ *
+ * @param {IncomingMessage || ReadableStream} streamIn
+ * @param {Response} streamOut
+ */
+async function getRequestedBlockFromCar(streamIn, streamOut) {
+  const carBlockIterator = await CarBlockIterator.fromIterable(streamIn);
+  let count = 0;
+
+  for await (const { cid, bytes } of carBlockIterator) {
+    if (!validateCarBlock(cid, bytes)) {
+      streamOut.status(502);
+      break;
+    }
+
+    // Is it safe to assume the first block is the requested one?
```

Comment:
Pretty safe for now but you have an easy opportunity here to check the cid against the requested one don't you? If you have the requested cid handy then pass it into here and check it (safest would be do a Buffer compare of the multihash bytes or up-convert both to cidv1 and string compare their base32 form). You could also do a check on the CarBlockReader's roots that there is just one and it matches what you asked for. Worth doing some sanity checking that things are working as expected. If this ever breaks, you should raise it with us (lassie). It's not an explicit contract but perhaps we should make it one.

Reply:
> Pretty safe for now but you have an easy opportunity here to check the cid against the requested one don't you?
ya for sure but I was held back by the possibility of a file path in the URL, in which case I don't know which cid to match against until Lassie returns CAR files that includes the parent CIDs.

I will add some sanity checks for when there's no path.

Reply:
I haven't followed in detail, but I think we're signed up for only verifiable CARs so even in the case of a path the first block should be the one requested. At least I think that's a reasonable thing to build in to a requirement.

Reply:
Lassie always returns parent CDs. I think you are good.